### PR TITLE
Change implementation of IndexUtilities to guarantee consistent indexing

### LINKED ISSUE #25: Error with IB point on a patch corner with IndexUtilities::getCellIndex()
Try for example tests/Stokes-IB/ex2 with 4 processors.

### PR DESCRIPTION
**Pull request by boyceg**
Add support for globalized indexing in IndexUtilities. It makes IndexUtilities a lightweight object that has an associated GridGeometry object that is used to compute unique Index values for a given position.

This PR replaces #38. It differs from that PR in that it avoids static global variables. This PR fixes #25.

### CODE REVIEW

**Comment by boyceg** [comment]
@amneetb, please review.

File: ibtk/src/math/PETScMatUtilities.cpp [hunk 1]
```diff
@@ -641,7 +641,7 @@ void PETScMatUtilities::constructPatchLevelSCInterpOp(Mat& mat,
     for (int k = 0; k < n_local_points; ++k)
     {
         const double* const X = &X_arr[NDIM * k];
-        const Index<NDIM> X_idx = IndexUtilities::getCellIndex(X, x_lower, x_upper, dx, domain_lower, domain_upper);
+        const Index<NDIM> X_idx = IndexUtilities::getCellIndexLocal(X, x_lower, x_upper, dx, domain_lower, domain_upper);
```
Comment (amneetb) [inline ibtk/src/math/PETScMatUtilities.cpp]: This for sure causes a bug in parallel, unless you use the global version of it.

**Comment by amneetb** [comment]
I would insist to change every call to the global version in the library. You can never be sure with the local version, e.g., I am pretty sure that local version will cause a bug in parallel with PETScMatUtilities::constructSCInterpOp(). This was whole reason I started the PR ;-)

**Comment by boyceg** [comment]
> On Jan 20, 2016, at 12:59 PM, Amneet Bhalla <notifications@github.com> wrote:
> 
> I would insist to change every call to the global version in the library. You can never be sure with the local version, e.g., I am pretty sure that local version will cause a bug in parallel with PETScMatUtilities::constructSCInterpOp(). This was whole reason I started the PR ;-)
> 
For cases that do not need to use the global version (e.g. they are performing operations where it does not matter which index is assigned), it is OK to use the local version.

File: ibtk/src/math/PETScMatUtilities.cpp (hunk 1 above)
Comment (amneetb) [inline ibtk/src/math/PETScMatUtilities.cpp]: Something like:

Pointer<CartesianGridGeometry<NDIM> > grid_geom = patch_level->getGridGeometry();
IndexUtilities indexer(grid_geom);
indexer.getCellIndexGlobal(...);

**Comment by boyceg** [comment]
I have switched over a few more cases from the local version to the global one. I think that the ones that remain local should be fine.

**Comment by boyceg** [comment]
I think there was one edge case left in the LMarkerRefine operator.

@amneetb please see if this fixes your parallel indexing problems.

**Comment by amneetb** [comment]
OK, I merged this PR to boyceg/MG-smoothers. Stokes-IB tests and standard ib examples are running fine.

**Comment by amneetb** [comment]
error with std:: on linux.

../../../../IBAMR/ibtk/include/ibtk/private/IndexUtilities-inl.h:227:20: error: no member named 'round' in namespace 'std'; did you mean
      simply 'round'?
        ratio(d) = std::round(dx0[d] / dx[d]);
                   ^~~~~~~~~~
                   round
/usr/include/x86_64-linux-gnu/bits/mathcalls.h:309:14: note: 'round' declared here
__MATHCALLX (round,, (_Mdouble_ __x), (__const__));

**Comment by boyceg** [comment]
> On Jan 20, 2016, at 4:16 PM, Amneet Bhalla <notifications@github.com> wrote:
> 
> error with std:: on linux.
> 
> ../../../../IBAMR/ibtk/include/ibtk/private/IndexUtilities-inl.h:227:20: error: no member named 'round' in namespace 'std'; did you mean
> simply 'round'?
> ratio(d) = std::round(dx0[d] / dx[d]);
> ^~~~~~~~~~
> round
> /usr/include/x86_64-linux-gnu/bits/mathcalls.h:309:14: note: 'round' declared here
> MATHCALLX (round,, (Mdouble __x), (__const));
> 
Whoops.

http://stackoverflow.com/questions/12696764/round-is-not-a-member-of-std <http://stackoverflow.com/questions/12696764/round-is-not-a-member-of-std>
http://stackoverflow.com/questions/485525/round-for-float-in-c/24348037#24348037 <http://stackoverflow.com/questions/485525/round-for-float-in-c/24348037#24348037>

Maybe we should use boost round:

http://www.boost.org/doc/libs/1_53_0/libs/math/doc/sf_and_dist/html/math_toolkit/utils/rounding/round.html <http://www.boost.org/doc/libs/1_53_0/libs/math/doc/sf_and_dist/html/math_toolkit/utils/rounding/round.html>

**Comment by boyceg** [comment]
@amneetb, I think this is really the final version. Adding state to IndexUtilities was dumb; I will just provide versions of getCellIndex that take in the patch or grid geometries.

**Comment by boyceg** [comment]
> On Jan 20, 2016, at 4:16 PM, Amneet Bhalla <notifications@github.com> wrote:
> 
> error with std:: on linux.
> 
> ../../../../IBAMR/ibtk/include/ibtk/private/IndexUtilities-inl.h:227:20: error: no member named 'round' in namespace 'std'; did you mean
> simply 'round'?
> ratio(d) = std::round(dx0[d] / dx[d]);
> ^~~~~~~~~~
> round
> /usr/include/x86_64-linux-gnu/bits/mathcalls.h:309:14: note: 'round' declared here
> MATHCALLX (round,, (Mdouble __x), (__const));
> 
This should be fixed with the latest commit --- I change all calls to round() to use the Boost round function. (I always forget that std::round() is not available until C++11.)

**Comment by boyceg** [comment]
@amneetb please check this revised version, which goes back to all-static functions for IndexUtilities.

**Comment by amneetb** [comment]
https://github.com/IBAMR/IBAMR/blob/73b92e945df22f108a1268d9bb265a2a63fd8dd9/ibtk/include/ibtk/private/IndexUtilities-inl.h#L117

This line looks suspect with reference.

On Wed, Jan 20, 2016 at 10:53 PM, Boyce Griffith <notifications@github.com>
wrote:

> @amneetb <https://github.com/amneetb> please check this revised version,
> which goes back to all-static functions for IndexUtilities.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/IBAMR/IBAMR/pull/39#issuecomment-173476570>.
>



-- 
--Amneet
